File: server-pedidos-java/src/main/java/br/edu/utfpr/serverpedidosjava/controller/ProdutoController.java
package br.edu.utfpr.serverpedidosjava.controller;

import br.edu.utfpr.serverpedidosjava.controller.dto.ProdutoDTO;
import br.edu.utfpr.serverpedidosjava.entity.Produto;
import br.edu.utfpr.serverpedidosjava.repository.ProdutoRepository;
import jakarta.annotation.Nullable;
import jakarta.validation.Valid;
import org.apache.logging.log4j.util.Strings;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/produtos")
public class ProdutoController {

    private final ProdutoRepository produtoRepository;

    public ProdutoController(ProdutoRepository produtoRepository) {
        this.produtoRepository = produtoRepository;
    }

    @GetMapping
    public List<ProdutoDTO> list(@RequestParam(name = "query") @Nullable String query) {
        List<Produto> produtos;
        if (Strings.isNotBlank(query)) {
            produtos = this.produtoRepository.findByQuery(query);
        } else {
            produtos = this.produtoRepository.findAll();
        }
        return produtos.stream().map(ProdutoDTO::fromEntity).toList();
    }

    @GetMapping("/{id}")
    public ResponseEntity<ProdutoDTO> findById(@PathVariable(name = "id") Integer id) {
        Produto produto = this.produtoRepository.findById(id).orElse(null);
        if (produto == null) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.ok(ProdutoDTO.fromEntity(produto));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> delete(@PathVariable(name = "id") Integer id) {
        this.produtoRepository.deleteById(id);
        return ResponseEntity.noContent().build();
    }

    @PostMapping
    public ProdutoDTO save(@Valid @RequestBody ProdutoDTO produto) {
        Produto produtoAtualizado = this.produtoRepository.save(produto.toEntity());
        return ProdutoDTO.fromEntity(produtoAtualizado);
    }
}

File: server-pedidos/src/main/kotlin/br/edu/utfpr/serverpedidos/controller/ProdutoController.kt
package br.edu.utfpr.serverpedidos.controller

import br.edu.utfpr.serverpedidos.entity.Produto
import br.edu.utfpr.serverpedidos.repository.ProdutoRepository
import jakarta.validation.Valid
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import kotlin.jvm.optionals.getOrNull

@RestController
@RequestMapping("/produtos")
class ProdutoController(
    private val produtoRepository: ProdutoRepository
) {

    @GetMapping
    fun list(@RequestParam query: String?): List<Produto> {
        query?.let {
            return produtoRepository.findByQuery(it)
        }
        return produtoRepository.findAll()
    }

    @GetMapping("/{id}")
    fun findById(@PathVariable(name = "id") codigo: Int): ResponseEntity<Produto> {
        val produto = produtoRepository.findById(codigo).getOrNull()
            ?: return ResponseEntity
                .notFound()
                .build()
        return ResponseEntity.ok(produto)
    }

    @DeleteMapping("/{id}")
    fun delete(@PathVariable id: Int): ResponseEntity<Unit> {
        produtoRepository.deleteById(id)
        return ResponseEntity
            .noContent()
            .build()
    }

    @PostMapping
    fun save(@Valid @RequestBody produto: Produto): Produto {
        return produtoRepository.save(produto)
    }
}
